internal/resource/validators/validator_utils.go

```diff
@@ -0,0 +1,135 @@
+}
+
+func SchemaUrlFormat() validator.String {
+	return stringvalidator.RegexMatches(regexp.MustCompile(`^(https?:\/\/)`), "Invalid entry for \"base_url\"! This value must start with 'http://' or 'https://'")
+}
+
+func SchemaHostnameOrIp() validator.String {
+	return stringvalidator.RegexMatches(regexp.MustCompile(`^([a-z0-9]+(-[a-z0-9]+)*\.)*[a-z0-9]+(-[a-z0-9]+)*$`), "Invalid hostname or IP, Please provide a valid hostname or IP address")
+}
+
+func SchemaEmail() validator.String {
+	return stringvalidator.RegexMatches(regexp.MustCompile(`^[a-zA-Z0-9._%+-]+@[a-zA-Z0-9.-]+\.[a-zA-Z]{2,}$`), "Invalid email address! Must be of the form '<address>@<company>.<domain>', where 'domain' contains only alphabetic characters and is at least 2 characters in length.")
+}
+
+func IsUrlFormat(aV attr.Value, validateConfigResp *resource.ValidateConfigResponse) diag.Diagnostics {
+	// ToLower on the MatchString call makes this case-insensitive
+	// Regex Breakdown
+	//^ and $: Start and end of the string.
+	// (https?:\/\/): Matches the http:// or https:// scheme. This part is optional
+	// ([\da-z\.-]+): Matches the domain name, which can contain digits, alphabets (lowercase), dots, and hyphens.
+	//       One or more such characters must be present.
+	// (\.[a-z]{2,6}(\.[a-z]{2,6})?)?: Matches the optional TLD and sub-TLD, each containing 2 to 6 alphabets (lowercase).
+	//       The entire TLD part is optional.
+	// (:[0-9]{2,5})?: Matches the optional port number, which can be 2 to 5 digits long.
+	// (\/[^\s\/]+[^\s]*)? matches the optional path
+	//
+	// The regex will match the following examples:
+	// 	http://example.com
+	//	https://example.com
+	//	http://example.com:8080
+	//	https://example.co.uk:8080
+	//	https://localhost
+	//	http://localhost:3000
+	//	http://example.com/path/to/resource
+	//  https://anotherexample.co.uk:9399/path/with/trailing/slash/
+	//  localhost
+	//  localhost:9999
+	//
+	//  Non-Matching examples
+	//	http://example. (TLD ends with a dot)
+	//	http://example.com:808080 (Port number exceeds 5 digits)
+	//	http://example_com.com (Underscore is not allowed in domain name)
+	//	ftp://example.com (Only http and https schemes are allowed if specified)
+	basetypesStringVal := aV.(types.String)
+	re, _ := regexp.Compile(`^(https?:\/\/)?([\da-z\.-]+)(\.[a-z]{2,6}(\.[a-z]{2,6})?)?(:[0-9]{2,5})?(\/)?(\/[^\s]*)?$`)
+	isUrl := re.MatchString(strings.ToLower(aV.(types.String).ValueString()))
+	if !internaltypes.IsNonEmptyString(basetypesStringVal) && !isUrl {
+		diag := validateConfigResp.Diagnostics
+		diag.AddError("Invalid URL Format!", fmt.Sprintf("Please provide a valid origin. Origin \"%s\" needs to be in a valid URL-like format - \"http(s)//:<value>.<domain>\"", basetypesStringVal.ValueString()))
+		return diag
+	}
+	return nil
+}
+
+func IsEmailFormat(aV attr.Value, validateConfigResp *resource.ValidateConfigResponse) diag.Diagnostics {
```

Comment:
Better to just return diagnostics here and in the next couple methods, no need to pass in the response object. The caller of this method can add to the response if necessary. Also, it's better to return an empty diagnostics objects rather than nil if there are no diagnostics to return.